build_runner_core/lib/src/logging/failure_reporter.dart

```diff
@@ -0,0 +1,117 @@
+import 'dart:io';
+
+import 'package:build/build.dart';
+import 'package:logging/logging.dart';
+import 'package:path/path.dart' as p;
+
+import '../asset_graph/node.dart';
+import '../util/constants.dart';
+
+/// A tracker for the errors which have already been reported during the current
+/// build.
+///
+/// Errors that occur due to actions that are run within this build will be
+/// reported directly as they happen. When an action is skipped and remains a
+/// failure the error will not have been reported by the time we check wether
+/// the build is failed.
+///
+/// The lifetime of this class should be a single build.
+class FailureReporter {
+  /// Removes all stored errors from previous builds.
+  ///
+  /// This should be called any time the build phases change since the naming
+  /// scheme is dependent on the build phases.
+  static Future<void> cleanErrorCache() async {
+    final errorCacheDirectory = new Directory(p.fromUri(errorCachePath));
+    if (await errorCacheDirectory.exists()) {
+      await errorCacheDirectory.delete(recursive: true);
+    }
+  }
+
+  /// Remove the stored error for [phaseNumber] runnon on [primaryInput].
+  ///
+  /// This should be called anytime the action is being run.
+  static Future<void> clean(int phaseNumber, AssetId primaryInput) async {
+    final errorFile =
+        new File(_errorPathForPrimaryInput(phaseNumber, primaryInput));
+    if (await errorFile.exists()) {
+      await errorFile.delete();
+    }
+  }
+
+  /// A set of Strings which uniquely identify a particular build action and
+  /// it's primary input.
+  final _reportedActions = new Set<String>();
+
+  /// Indicate that a failure reason for the build step which would produce
+  /// [output] and all other outputs from the same build step has been printed.
+  Future<void> markReported(String actionDescription, GeneratedAssetNode output,
+      Iterable<ErrorReport> errors) async {
+    if (!_reportedActions.add(_actionKey(output))) return;
+    final errorFile =
+        await new File(_errorPathForOutput(output)).create(recursive: true);
+    await errorFile.writeAsString(jsonEncode(<dynamic>[actionDescription]
+        .followedBy(errors
+            .map((e) => [e.message, e.error, e.stackTrace?.toString() ?? '']))
+        .toList()));
+  }
+
+  /// Indicate that the build steps which would produce [outputs] are failing
+  /// due to a dependency and being skipped so no actuall error will be
+  /// produced.
+  Future<void> markSkipped(Iterable<GeneratedAssetNode> outputs) =>
+      Future.wait(outputs.map((output) async {
+        if (!_reportedActions.add(_actionKey(output))) return;
+        await clean(output.phaseNumber, output.primaryInput);
+      }));
+
+  /// Log stored errors for any build steps which would output nodes in
+  /// [failingNodes] which haven't already been reported.
+  Future<void> reportErrors(Iterable<GeneratedAssetNode> failingNodes) async {
+    for (final failure in failingNodes) {
```

Comment:
it is probably worth parallelizing this

Reply:
Done. Do you think I should add a file handle pool? This isn't really in the critical path for a build so it might not make much difference.

Reply:
hmm, you could just use the asset reader as well which would handle this? If somebody has a lot of cascading errors it could cause problems.

Reply:
I'd prefer to leave it using `dart:io` for now, for consistency.

Since we do directory level work we can't use asset reader/writer for everything.